Add tool to check local materials against products of passed link

lukpueh opened the pull request:
Adds new runlib functionary API, to perform basic vetting of materials. More specifically, to check if the **untrusted** materials match the products of a passed **trusted** link metadata file.

**Caveat**
The used link metadata is not validated against a layout. This means, no checking if the link is part of a certain supply chain, or if it was signed by a threshold of authorised functionaries.

--- 

This PR includes: API, CLI, RTD docs, tests and an small unrelated docs fix. Review commit by commit for details.

Commit:
runlib: add basic local material checker

Adds new runlib functionary API, to perform basic vetting of materials.
More specifically, to check if the **untrusted** materials match the
products of a passed **trusted** link metadata file.

**Caveat**
The used link metadata is not validated against a layout. This means, no
checking if the link is part of a certain supply chain, or if it was
signed by a threshold of authorized functionaries.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
cli: for local material checker

- Loads link metadata from passed path
- Calls runlib API, to record local materials and match with products
- Returns 0 if they match, and 1 otherwise.
- Optionally provides verbose output.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
doc: add missing blank line to fix doc build

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
doc: add rtd docs for check material api+cli

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

lukpueh commented:
kudos to @adityasaky for starting this feature in #582 🎉

adityasaky commented on in_toto/in_toto_check_materials.py:
```diff
@@ -0,0 +1,105 @@
+                print(f"Only in products: {name}")
+
+            for name in only_materials:
+                print(f"Only in materials: {name}")
```
Can we say something other than `materials`? Depending on context of use, it may not fir the `materials` terminology, say at the last step.

lukpueh replied:
Sure. Although, in the last step we'd still be checking the materials of the last step, right? I think the terminology fits the use case that motivated this. But you are right that the tool could be used for something else too. Let me adopt the naming you suggested in the original PR.

adityasaky replied:
Apologies, I was thinking of how we'd align it with final product verification if we wanted to in future.

lukpueh replied:
I renamed in the last commit. Please re-approve.

adityasaky reviewed (approved):
Very neat and clean! Thanks for picking this one up. :smile:

Commit:
rename: check-materials -> match-products

Although the use case that motivated this tool, was to check untrusted
materials. The tool may also be used to check artifacts that are not
materials. In either case, the local artifacts are matched against the
products of a link. Hence the new name. h/t @adityasaky

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

adityasaky commented on tests/test_runlib.py:
```diff
@@ -1042,5 +1043,85 @@ def test_create_metadata_with_command_byproducts_environment(self):
     os.remove(self.link_name)
 
+
+class InTotoMatchProducts(TmpDirMixin, unittest.TestCase):
```
```suggestion
class TestInTotoMatchProducts(TmpDirMixin, unittest.TestCase):
```

nit: consistency

adityasaky commented on tests/test_in_toto_match_products.py:
```diff
@@ -0,0 +1,82 @@
+from in_toto.models.metadata import Metadata
+
+
+class InTotoMatchProducts(unittest.TestCase):
```
```suggestion
class TestInTotoMatchProducts(unittest.TestCase):
```

nit: consistency

lukpueh committed:
Apply suggestions from code review

Co-authored-by: Aditya Sirish <8928778+adityasaky@users.noreply.github.com>
Signed-off-by: Lukas Pühringer <luk.puehringer@gmail.com>

adityasaky commented:
Thanks @lukpueh!
